### PR TITLE
interface'ifying things

### PR DESCRIPTION
**Pull request by frrist**
# What's in the PR?
- Core interface changes
- Plugin system (localipfs & dockeripfs)
    - The plugin implementation is baebones, improvements to the *ipfs plugins are desirable.
- Reworked CLI
- Increased Test Coverage

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -74,18 +72,17 @@ func main() {
...
-		connectCmd,
-		dumpStacksCmd,
-		forEachCmd,
-		getCmd,
 		initCmd,
+
+		startCmd,
 		killCmd,
 		restartCmd,
-		setCmd,
```
Comment (whyrusleeping) [inline main.go]: where is set going?

**Comment by whyrusleeping** [comment]
@travisperson could you take a look at incorporating these changes too while youre at it? https://github.com/ipfs/iptb/pull/60/commits/08f0993a0e00bbb62266b63fd3f4366739d39242

It separates the init process into two steps (optionally) then makes the normal init process explicitly do both. This will make certain setups easier to play with in the future.

**Comment by travisperson** [comment]
@whyrusleeping ah nice. Ya should be easy to incorporate.

File: testbed/spec.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+func loadPlugins() {
+	dir, err := pluginDir()
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	plugs, err := ioutil.ReadDir(dir)
+
```
Comment (whyrusleeping) [inline testbed/spec.go]: style nit: don't put spaces between creating and checking an error

File: testbed/spec.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	NewNode := *(NewNodeSym.(*testbedi.NewNodeFunc))
+
+	PluginNameSym, err := pl.Lookup("PluginName")
+	if err != nil {
+		return err
+	}
+
+	PluginName := *(PluginNameSym.(*string))
+
+	plugins[PluginName] = iptbplugin{
```
Comment (whyrusleeping) [inline testbed/spec.go]: probably want to check if a plugin by this name exists already

**Comment by whyrusleeping** [comment]
We should make sure its possible to build iptb by default with support for some 'default plugins'. I want to not have to worry about where my ipfs plugin is located

**Comment by travisperson** [comment]
Building in some plugins is a great idea, I updated the current "plugins" to get built into the binary and allow for them to be overridden.

Also exposed them as a plugin package for easy hacking / building your own :hammer:

File: plugins/ipfs/util.go [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+	}
+
+	return pcid, nil
+}
+
+func GetAttrList() []string {
+	return []string{attrId, attrPath, attrBwIn, attrBwOut}
+}
+
+func GetAttrDesc(attr string) (string, error) {
```
Comment (whyrusleeping) [inline plugins/ipfs/util.go]: duplicate code?

File: testbed/interfaces/node.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	"github.com/ipfs/go-cid"
+	"github.com/multiformats/go-multiaddr"
+)
+
+type NewNodeFunc func(dir string, extras map[string]interface{}) (TestbedNode, error)
+type GetAttrListFunc func() []string
+type GetAttrDescFunc func(attr string) (string, error)
+
+// TestbedNode specifies the interface to a process controlled by iptb
+type TestbedNode interface {
```
Comment (whyrusleeping) [inline testbed/interfaces/node.go]: This interface is getting pretty big, I wonder if we should break it up, and make some of the methods optional (by specialization, or similar)

File: main.go [hunk 6]
```diff
@@ -500,16 +717,21 @@ var runCmd = cli.Command{
...
-		nd, err := util.LoadNodeN(n)
+		tb := testbed.NewTestbed(testbeddir)
+
+		nd, err := tb.Node(n)
 		if err != nil {
 			return err
 		}
 
-		out, err := nd.RunCmd(c.Args()[1:]...)
+		out, err := nd.RunCmd(context.TODO(), c.Args()[1:]...)
```
Comment (frrist) [inline main.go]: can pass a context with a timeout here

File: plugins/ipfs/docker/dockeripfs.go [hunk 7]
```diff
@@ -0,0 +1,547 @@
...
+var NewNode testbedi.NewNodeFunc
+var GetAttrDesc testbedi.GetAttrDescFunc
+var GetAttrList testbedi.GetAttrListFunc
+
+const (
+	attrIfName = "ifname"
+)
+
+func init() {
+	NewNode = func(dir string, extras map[string]interface{}) (testbedi.TestbedNode, error) {
```
Comment (whyrusleeping) [inline plugins/ipfs/docker/dockeripfs.go]: You don't have to assign these like this, it should work the same if you just define it as:
```
func NewNode(dir string, extras map[string]interface{}) (testbedi.TestbedNode, error) {
...
...
}
```

File: plugins/ipfs/docker/dockeripfs.go [hunk 8]
```diff
@@ -0,0 +1,547 @@
...
+	alive, err := l.isAlive()
+	if err != nil {
+		return err
+	}
+
+	if alive {
+		return fmt.Errorf("node is already running")
+	}
+
+	cmd := exec.Command("docker", "run", "-d", "-v", l.dir+":/data/ipfs", l.image)
```
Comment (whyrusleeping) [inline plugins/ipfs/docker/dockeripfs.go]: Interesting future work here would be to use the docker api package: https://docs.docker.com/develop/sdk/#sdk-and-api-quickstart

File: testbed/interfaces/node.go (hunk 5 above)
Comment (frrist) [reply testbed/interfaces/node.go]: The return types and parameters are a little off, but a rough idea:
https://gist.github.com/frrist/01405d6b36aae85108fad80c2c40d3fb

**Comment by whyrusleeping** [comment]
Before merging, let's make sure that the tests all pass, and that if we use this in the go-ipfs sharness tests, they all still pass

File: commands/attr.go [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+		argValue := c.Args()[2]
+
+		i, err := strconv.Atoi(argNode)
+		if err != nil {
+			return fmt.Errorf("parse err: %s", err)
+		}
+
+		tb := testbed.NewTestbed(path.Join(flagRoot, "benches", flagTestbed))
+
+		if flagSave {
```
Comment (frrist) [inline commands/attr.go]: Should we first attempt to set the attribute on the node before persisting it to the nodespec file?

File: commands/fast.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+			return err
+		}
+
+		var list []int
+		for i, _ := range nodes {
+			list = append(list, i)
+		}
+
+		runCmd := func(node testbedi.Core) (testbedi.TBOutput, error) {
+			return node.Init(context.TODO(), ([]string{})...)
```
Comment (frrist) [inline commands/fast.go]: can remove `([]string{})...`

File: commands/start.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+			Name:  "wait",
+			Usage: "wait for nodes to start before returning",
+		},
+		cli.BoolFlag{
+			Name:   "terminator",
+			Hidden: true,
+		},
+	},
+	Before: func(c *cli.Context) error {
+		if present := isTerminatorPresent(c); present {
```
Comment (frrist) [inline commands/start.go]: let's leave a comment describing what `terminator` means

File: commands/utils.go [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+			defer wg.Done()
+			out, err := fn(node)
+
+			lk.Lock()
+			defer lk.Unlock()
+
+			results[i] = Result{
+				Node:   n,
+				Output: out,
+				Error:  errors.Wrapf(err, "node[%d]", n),
```
Comment (frrist) [inline commands/utils.go]: will this this be incorrect if `list` is a range like [4-8]?

Comment (travisperson) [reply commands/utils.go]: yep!

File: commands/start.go (hunk 11 above)
Comment (travisperson) [reply commands/start.go]: I added a comment on `isTerminatorPresent` as this is used on any command which takes a node range.

**Comment by travisperson** [comment]
To make sure everything was working correctly I did a quick update to go-ipfs sharness tests as well using the new cli.

You can see those changes here https://github.com/ipfs/go-ipfs/compare/master...travisperson:update-iptb?expand=1

File: plugins/ipfs/util.go [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+
+	return resp.Body, nil
+}
+
+type BW struct {
+	TotalIn  int
+	TotalOut int
+}
+
+func GetBW(l testbedi.Libp2p) (*BW, error) {
```
Comment (Stebalien) [inline plugins/ipfs/util.go]: Might as well take a context for all of these functions up-front (unless I'm missing something about how they're supposed to be used).

File: plugins/ipfs/util.go [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+	return string(out), err
+}
+
+func SwarmAddrs(l testbedi.Core) ([]string, error) {
+	pcid, err := l.PeerID()
+	if err != nil {
+		return nil, err
+	}
+
+	output, err := l.RunCmd(context.TODO(), nil, "ipfs", "swarm", "addrs", "local")
```
Comment (Stebalien) [inline plugins/ipfs/util.go]: Again, taking a context in `SwarmAddrs` would allow us to feed it into this function here.

File: testbed/interfaces/node.go [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+// Attributes returned should be queriable through the Attribute interface.
+// Examples include: api address, peerid, cpu/ram limits, jitter.
+type GetAttrListFunc func() []string
+
+// GetAttrDescFunc returns the description of the attribute `attr`
+type GetAttrDescFunc func(attr string) (string, error)
+
+type Libp2p interface {
+	// PeerID returns the peer id
+	PeerID() (string, error)
```
Comment (Stebalien) [inline testbed/interfaces/node.go]: Any reason not to use actual peer IDs here instead of strings?

File: testbed/interfaces/node.go [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+	// APIAddr returns the multiaddr for the api
+	APIAddr() (string, error)
```
Comment (Stebalien) [inline testbed/interfaces/node.go]: Could use a multiaddr.

File: testbed/interfaces/node.go [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+	// SwarmAddrs returns the swarm addrs for the node
+	SwarmAddrs() ([]string, error)
```
Comment (Stebalien) [inline testbed/interfaces/node.go]: Should probably use multiaddrs.

File: testbed/interfaces/node.go [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+type Config interface {
+	Core
+	// GetConfig returns the configuration of the node
+	GetConfig() (interface{}, error)
+	// WriteConfig writes the configuration of the node
+	WriteConfig(interface{}) error
```
Comment (Stebalien) [inline testbed/interfaces/node.go]: Interfaces like this always make me a bit nervous. Be careful about sharing and race conditions. If this just writes a "json-like" thing, consider using refmt to copy it.

File: testbed/interfaces/node.go [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+// Attributes are ways to shape process execution and additional information that alters the
+// environment the process executes in
+type Attribute interface {
```
Comment (Stebalien) [inline testbed/interfaces/node.go]: Is this an "Attribute" or "Attribute*s*"?

File: testbed/interfaces/node.go (hunk 18 above)
Comment (Stebalien) [reply testbed/interfaces/node.go]: Actually, what is the "form" of the config? I.e., what types of strings, what types of values, etc.

File: testbed/interfaces/node.go [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+	   - Limit
+
+	   * RAM
+	   - Limit
+
+	*/
+}
+
+// Metrics are ways to gather information during process execution
+type Metric interface {
```
Comment (Stebalien) [inline testbed/interfaces/node.go]: Metric**s**?

File: testbed/testbed.go [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+func (tb *BasicTestbed) loadNodes() ([]testbedi.Core, error) {
+	specs, err := tb.Specs()
+	if err != nil {
+		return nil, err
+	}
+
+	return NodesFromSpecs(specs)
+}
+
+func InitNodes(nodes []testbedi.Core, args ...string) error {
```
Comment (Stebalien) [inline testbed/testbed.go]: I'd just pass in the context.

File: testbed/spec.go [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+
+	if plg, ok := plugins[pluginName]; ok {
+		return plg.NewNode(ns.Dir, ns.Attrs)
+	}
+
+	return nil, fmt.Errorf("Could not find plugin %s", pluginName)
+}
+
+// SetAttr sets an attribute on the NodeSpec
+func (ns *NodeSpec) SetAttr(attr string, val interface{}) {
```
Comment (Stebalien) [inline testbed/spec.go]: The `Attribute` interface takes/returns a string value. I assume these should match.

File: plugins/ipfs/docker/dockeripfs.go [hunk 23]
```diff
@@ -0,0 +1,611 @@
...
+	return ipfs.GetMetricList()
+}
+
+func GetMetricDesc(attr string) (string, error) {
+	return ipfs.GetMetricDesc(attr)
+}
+
+/// Core Interface
+
+func (l *Dockeripfs) Init(ctx context.Context, agrs ...string) (testbedi.Output, error) {
```
Comment (Stebalien) [inline plugins/ipfs/docker/dockeripfs.go]: s/agrs/args.

File: plugins/ipfs/util.go [hunk 24]
```diff
@@ -0,0 +1,310 @@
...
+	LT_TCP ListenType = iota
+	LT_WS
+	LT_UTP
+)
+
+func InitIpfs(l testbedi.Core) error {
+	return nil
+}
+
+func SwarmAddr(lt ListenType, addr string, port int) string {
```
Comment (Stebalien) [inline plugins/ipfs/util.go]: We should really just use normal multiaddrs. This'll be a pain to maintain.

File: testbed/interfaces/node.go [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+
+	*/
+}
+
+// Core specifies the interface to a process controlled by iptb
+type Core interface {
+	Libp2p
+	// Allows a node to run any initialization it may require
+	// Ex: Installing additional dependencies / setting up configuration
+	Init(ctx context.Context, args ...string) (Output, error)
```
Comment (Stebalien) [inline testbed/interfaces/node.go]: I assume that "args" are additional command line arguments, right?

File: README.md [hunk 26]
```diff
@@ -32,41 +29,57 @@ hey!
...
-	start			start up all testbed nodes
-	kill, stop		kill a specific node (or all nodes, if none specified)
-	restart			kill all nodes, then restart
-	shell			spawn a subshell with certain IPFS environment variables set
-	get			get an attribute of the given node
-	connect			connect two nodes together
-	dump-stack		get a stack dump from the given daemon
-	help, h			show a list of subcommands, or help for a specific subcommand
+     fast     create default bench and initialize
+     bench    create, remove, list bench setups
```
[the remark this answers is not in the document]
Comment (travisperson) [reply README.md]: Ya, not benchmarks, just different setups. I'll try to make that clear. There is some discussion around the introduction of `bench` in issue #62. Basically this stemmed from having both a `init` and a `create` command, and trying to make it less ambiguous as to which to run.

The thought was to place the concept of "create" under a category, and a suggestion was made to call them `benches`, like having a toolbench where you would work on something.

In the current implementation `init` no longer makes the node specification. Instead it's a direct map to the `Init` method on the node interface. This allows iptb to have it's own configuration (defining number of nodes, types, etc), and still exposing the direct call to the interface.

File: README.md [hunk 27]
```diff
@@ -32,41 +29,57 @@ hey!
...
-	init			create and initialize testbed configuration
-	start			start up all testbed nodes
-	kill, stop		kill a specific node (or all nodes, if none specified)
-	restart			kill all nodes, then restart
-	shell			spawn a subshell with certain IPFS environment variables set
-	get			get an attribute of the given node
-	connect			connect two nodes together
-	dump-stack		get a stack dump from the given daemon
-	help, h			show a list of subcommands, or help for a specific subcommand
+     fast     create default bench and initialize
```
[the remark this answers is not in the document]
Comment (travisperson) [reply README.md]: Right now there is `iptb bench create [other flags] --init`. The `fast` command was added (name mostly a placeholder, though `auto` might be a good choice) to bring back the functionality of `iptb init`, as currently it maps directly to the node interface method `Init`. It would be nice to keep that exposed as a directly way to invoke that method without overloading it too much. This could also be fixed by changing out the `Core#Init` method name something else, and keeping `init` as the quick and easy way to get started with iptb.

Basically `iptb fast` is almost a direct mapping to `iptb bench create --init`, it just doesn't accept the idea of additional configuration through `--attr` flags.

File: commands/connect.go [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+	cli "github.com/urfave/cli"
+
+	"github.com/ipfs/iptb/testbed"
+)
+
+var ConnectCmd = cli.Command{
+	Category:  "CORE",
+	Name:      "connect",
+	Usage:     "connect nodes together",
+	ArgsUsage: "<nodes> <nodes>",
```
[the remark this answers is not in the document]
Comment (travisperson) [reply commands/connect.go]: Nope, nothing like that yet. Sounds like a good addition though :+1:

File: testbed/interfaces/node.go (hunk 17 above)
Comment (travisperson) [reply testbed/interfaces/node.go]: Going to answer these three here. Originally we had both cids and multiaddrs on the interface. Ended up stripping both of them primarily due to an issue around multiaddrs (though maybe it's not really an issue).

Basically, because the primary use of the interface is for plugins. If a plugin uses a multiaddr codec that the loader doesn't have, panics are bound to occur.

https://github.com/multiformats/go-multiaddr/blob/master/multiaddr.go#L64
https://github.com/multiformats/go-multiaddr/blob/master/multiaddr.go#L84

This isn't so much an issue with iptb, as it doesn't directly use this interface at the moment (used primarily inside of `connect`).

Maybe the panics are fine, and it up to the consumer to recover from them, I'm open to moving back as long as we thing it's better.

File: testbed/interfaces/node.go (hunk 18 above)
Comment (travisperson) [reply testbed/interfaces/node.go]: I guess it's kind of completely up to the plugin to decide. Maybe these don't make a lot of sense anymore here. They would primarily be use when consuming iptb as a package though, allowing the user to get the config and cast it themselves.

https://github.com/ipfs/iptb/pull/61/files#diff-c348f01f34fe135d26d91d3176809188R106

File: testbed/interfaces/node.go (hunk 19 above)
Comment (travisperson) [reply testbed/interfaces/node.go]: It's just an extension of the Core interface, so not a singular or plural. The interfaces were split to reduce the required amount of code to implement a plugin. Example, if a node does not implement the `Attribute` interface, when running commands such as `iptb attr get`, an error will be returned stating that the operation can't be done because the plugin does not support it.

File: testbed/interfaces/node.go (hunk 25 above)
Comment (travisperson) [reply testbed/interfaces/node.go]: Correct

File: plugins/ipfs/local/localipfs.go [hunk 29]
```diff
@@ -0,0 +1,504 @@
...
+	"github.com/ipfs/iptb/util"
+	"github.com/multiformats/go-multiaddr"
+	"github.com/pkg/errors"
+)
+
+var errTimeout = errors.New("timeout")
+
+var PluginName = "localipfs"
+
+type Localipfs struct {
```
Comment (dignifiedquire) [inline plugins/ipfs/local/localipfs.go]: nit: `LocalIpfs` and `DockerIpfs` seem to be more inline with other naming I have seen, especially as they are not a single word, but rather `Local IPFS` and `Docker IPFS` if you write them not as identifiers

File: plugins/ipfs/local/localipfs.go [hunk 30]
```diff
@@ -0,0 +1,504 @@
...
+	}
+
+	p, err := os.FindProcess(pid)
+	if err != nil {
+		return fmt.Errorf("error killing daemon %s: %s", l.dir, err)
+	}
+
+	waitch := make(chan struct{}, 1)
+	go func() {
+		p.Wait() //TODO: pass return state
```
Comment (dignifiedquire) [inline plugins/ipfs/local/localipfs.go]: should this be fixed before merging?

File: plugins/ipfs/local/localipfs.go [hunk 31]
```diff
@@ -0,0 +1,504 @@
...
+		return err
+	}
+
+	_, err = l.RunCmd(ctx, nil, "ipfs", "swarm", "connect", swarmaddrs[0])
+
+	return err
+}
+
+func (l *Localipfs) Shell(ctx context.Context, nodes []testbedi.Core) error {
+	shell := os.Getenv("SHELL")
```
Comment (dignifiedquire) [inline plugins/ipfs/local/localipfs.go]: would be nice if these env variables are global constants, such that other plugins could reuse them

File: testbed/interfaces/node.go [hunk 32]
```diff
@@ -0,0 +1,145 @@
...
+	// WriteConfig writes the configuration of the node
+	WriteConfig(interface{}) error
+}
+
+// Attributes are ways to shape process execution and additional information that alters the
+// environment the process executes in
+type Attribute interface {
+	Core
+	// GetAttr returns the value of attr
+	GetAttr(attr string) (string, error)
```
Comment (dignifiedquire) [inline testbed/interfaces/node.go]: This is inconsistent, with both go best practices and the metric interface, and should be just `Attr` and `SetAttr`.

File: testbed/interfaces/node.go [hunk 33]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+// Core specifies the interface to a process controlled by iptb
+type Core interface {
+	Libp2p
+	// Allows a node to run any initialization it may require
+	// Ex: Installing additional dependencies / setting up configuration
+	Init(ctx context.Context, args ...string) (Output, error)
+	// Starts the node
+	Start(ctx context.Context, wait bool, args ...string) (Output, error)
```
Comment (dignifiedquire) [inline testbed/interfaces/node.go]: missing explanation of what `wait` does

File: testbed/interfaces/node.go [hunk 34]
```diff
@@ -0,0 +1,145 @@
...
+	Stop(ctx context.Context, wait bool) error
+	// Runs a command in the context of the node
+	RunCmd(ctx context.Context, stdin io.Reader, args ...string) (Output, error)
+	// Connect the node to another
+	Connect(ctx context.Context, n Core) error
+	// Starts a shell in the context of the node
+	Shell(ctx context.Context, ns []Core) error
+
+	// Writes a log line to stdout
+	Infof(format string, args ...interface{})
```
Comment (dignifiedquire) [inline testbed/interfaces/node.go]: why does a plugin need to implement this?

File: testbed/interfaces/node.go [hunk 35]
```diff
@@ -0,0 +1,145 @@
...
+	// Writes a log line to stderr
+	Errorf(format string, args ...interface{})
+
+	// Dir returns the iptb directory assigned to the node
+	Dir() string
+	// Type returns a string that identifies the implementation
+	// Examples localipfs, dockeripfs, etc.
+	Type() string
+
+	String() string
```
Comment (dignifiedquire) [inline testbed/interfaces/node.go]: what is  `String`  supposed to do/where is it useful?

File: commands/utils.go [hunk 36]
```diff
@@ -0,0 +1,207 @@
...
+
+	if arguments[0] == "--" {
+		arguments = arguments[1:]
+	}
+
+	return args[0], arguments
+}
+
+func parseRange(s string) ([]int, error) {
+	if strings.HasPrefix(s, "[") && strings.HasSuffix(s, "]") {
```
[the remark this answers is not in the document]
Comment (whyrusleeping) [reply commands/utils.go]: this code wasnt changed in this PR. Any changes to it should be made in a followup

File: commands/utils.go [hunk 37]
```diff
@@ -0,0 +1,207 @@
...
+	arguments := args[1:]
+
+	if arguments[0] == "--" {
+		arguments = arguments[1:]
+	}
+
+	return args[0], arguments
+}
+
+func parseRange(s string) ([]int, error) {
```
[the remark this answers is not in the document]
Comment (whyrusleeping) [reply commands/utils.go]: same as above

File: commands/utils_test.go [hunk 38]
```diff
@@ -0,0 +1,105 @@
...
+func expect(t *testing.T, a interface{}, b interface{}) {
+	_, fn, line, _ := runtime.Caller(1)
+	fn = strings.Replace(fn, wd+"/", "", -1)
+
+	if !reflect.DeepEqual(a, b) {
+		t.Errorf("(%s:%d) Expected %v (type %v) - Got %v (type %v)", fn, line, b, reflect.TypeOf(b), a, reflect.TypeOf(a))
+	}
+}
+
+func TestParseRange(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (whyrusleeping) [reply commands/utils_test.go]: same as above

File: plugins/ipfs/docker/dockeripfs.go [hunk 39]
```diff
@@ -0,0 +1,611 @@
...
+}
+
+func NewNode(dir string, attrs map[string]interface{}) (testbedi.Core, error) {
+	imagename := "ipfs/go-ipfs"
+	lt := ipfs.LT_TCP
+	mdns := false
+
+	var repobuilder string
+
+	if v, ok := attrs["image"]; ok {
```
[the remark this answers is not in the document]
Comment (whyrusleeping) [reply plugins/ipfs/docker/dockeripfs.go]: sometimes thats nice, but often it ends up leading to less readable code. I'd prefer keeping the strings here.

File: plugins/ipfs/local/localipfs.go (hunk 29 above)
Comment (whyrusleeping) [reply plugins/ipfs/local/localipfs.go]: Yeah, `LocalIpfs` seems better.

File: testbed/interfaces/node.go (hunk 34 above)
Comment (whyrusleeping) [reply testbed/interfaces/node.go]: Yeah, I was thinking that this one could probably be left off.

File: testbed/interfaces/node.go (hunk 35 above)
Comment (whyrusleeping) [reply testbed/interfaces/node.go]: String is a builtin go interface. It defines how this object gets printed if someone passes it to a `Printf` type function

File: testbed/interfaces/node.go (hunk 15 above)
Comment (frrist) [reply testbed/interfaces/node.go]: I believe the reasoning was to make dealing with dependencies easier, e.g. 2 different plugins exists, one uses `go-cid` version foo, another plugin uses `go-cid` version bar. This means iptb will need to be built specifically with the version of `go-cid` the plugin requires, which is annoying. I believe the same holds true for multiformats.

File: testbed/interfaces/node.go (hunk 34 above)
Comment (travisperson) [reply testbed/interfaces/node.go]: Ya, I think we talked about this previously, it can be dropped. Original thought was to make it easier to use the interface in other applications (embedded iptb), and provide logging.

File: README.md [hunk 40]
```diff
@@ -32,41 +29,57 @@ hey!
...
-	--help, -h		show help
-	--version, -v		print the version
+   --bench value  Name of bench to use under IPTB_ROOT (default: "default") [$IPTB_BENCH]
+   --help, -h     show help
+   --version, -v  print the version
 ```
 
 ### Install
 
+`iptb` uses [gx](https://github.com/whyrusleeping/gx)
```
[the remark this answers is not in the document]
Comment (travisperson) [reply README.md]: We can do that as long as we drop the `ipfs` plugins from being built into the binary by default due to its heavy use of gx. Currently the `go-libp2p` package needs to be versioned.

File: testbed/testbed.go (hunk 21 above)
Comment (travisperson) [reply testbed/testbed.go]: dead code actually, deleting...

File: testbed/interfaces/node.go (hunk 33 above)
Comment (travisperson) [reply testbed/interfaces/node.go]: I think it needs to be removed actually. At the moment it doesn't serve much of a purpose. We can add it back if we find that it's actually needed for something.

Comment (travisperson) [reply testbed/interfaces/node.go]: Actually, I guess start has a purpose still. It can be to wait for the node to be reachable (api online). Stop is the one that doesn't really fit at the moment. It was added per a mention in #10, but I can't seem to think why it would be needed.

File: commands/connect.go [hunk 41]
```diff
@@ -0,0 +1,76 @@
...
+		nodes, err := tb.Nodes()
+		if err != nil {
+			return err
+		}
+
+		var results []Result
+		for _, f := range from {
+			for _, t := range to {
+				ctx, cancel := context.WithTimeout(context.Background(), timeout)
+				defer cancel()
```
[the remark this answers is not in the document]
Comment (frrist) [reply commands/connect.go]: Agreed, but this can happen in a follow on PR

File: commands/connect.go [hunk 42]
```diff
@@ -0,0 +1,76 @@
...
+	"github.com/pkg/errors"
+	cli "github.com/urfave/cli"
+
+	"github.com/ipfs/iptb/testbed"
+)
+
+var ConnectCmd = cli.Command{
+	Category:  "CORE",
+	Name:      "connect",
+	Usage:     "connect nodes together",
```
[the remark this answers is not in the document]
Comment (frrist) [reply commands/connect.go]: Agreed, I propose we address this in a follow on PR since similar changes will likely be required.

File: plugins/ipfs/local/localipfs.go (hunk 30 above)
Comment (travisperson) [reply plugins/ipfs/local/localipfs.go]: I believe this was copied over from some other code. Not sure what the original intent was. I've removed the todo.

File: sharness/t0020-stop.sh [hunk 43]
```diff
@@ -4,23 +4,24 @@ test_description="iptb stop tests"
...
 '
 
 test_expect_success "iptb start works" '
-	../bin/iptb start --args --debug
+	../bin/iptb start --wait -- --debug
 '
 
 test_expect_success "iptb stop works" '
-	../bin/iptb stop
+	../bin/iptb stop && sleep 1
```
Comment (whyrusleeping) [inline sharness/t0020-stop.sh]: we *could* have iptb sit around and wait for the processes to be really dead

Comment (whyrusleeping) [reply sharness/t0020-stop.sh]: at least make an issue for it, fine to fix later

**Comment by travisperson** [comment]
@Stebalien @dignifiedquire @whyrusleeping 

Is there any other feedback that should block this PR? There is additional work I want to do to the plugins. I can remove that code from this PR, or simply follow up.
